### DISCUSSION TITLE
passing initial solution or warm start

### DISCUSSION THREAD

**Discussion by LucaVaresi** [ERGO-Code/HiGHS#1401] [Q&A]
Hi everyone,
I have a question regarding HiGHS and in particular highspy.
I have looked online but i have not found a way to specify an initial solution to the HiGHS solver is there a way to do that?

**Comment by jajhall** [comment]
It's possible to do it using `Highs::setSolution`, but the binding for this in `highspy` was forgotten in the last release. It will get into `highspy` soon - possibly this week as we address another issue

**Reply by LucaVaresi** [reply]
thank you!
Just another question on this topic, what type of object should be passed to Highs::setSolution?

**Reply by jajhall** [reply]
An instance of the `HighsSolution` struct with (at least) values for the variables - `col_value`. These must correspond to a feasible solution to be useful. 

If you're solving a MIP, then (even) a partial assignment of the discrete variables will be exploited, so long as it corresponds to a feasible solution

**Comment by jajhall** [comment]
With https://pypi.org/project/highspy/1.7.1.dev1/ `Highs::setSolution` can now be used

**Reply by guifcoelho** [reply]
Hello @jajhall, could you provide some tips on how to pass a partial solution with the latest version of highspy?

I've managed to provide the whole solution (warmstarting all variables):

```
sol = highspy.HighsSolution()
sol.col_value = [var_value for var_value in variables]
model.setSolution(sol)
```

If all variables are binary and some of them are not set (with value None), I could do this:

```
sol = highspy.HighsSolution()
sol.col_value = [var_value or 0 for var_value in variables]
model.setSolution(sol)
```

But , what if I have some continuous or integer variables that I don't want to warm start? I've tried something like below and it doesn't work:

```
sol = highspy.HighsSolution()
sol.col_value = [1, 1, 1, None, None, 0, 1, None]
model.setSolution(sol)
```

**Reply by jajhall** [reply]
What value is `None`? 

It should be sufficient to set to (say) 0.5 integer/binary variables that you don't want to warm start.

I'll be working on #1692 in the coming weeks, if you have any more experience to share.

**Reply by guifcoelho** [reply]
What if my integer variable bounds are 5 and 50? Would it be sufficient to put any value in the middle?

I believe `None` is the equivalent of `NULL` in C++ (sorry if I'm wrong, it's been centuries since the last time I programmed in C++). I was hoping that filling `col_value` with some `None`s it would translate to `NULL`s to represent "empty" positions in the array. In my opinion, providing some way to say that this or that variable is not being set for warmstart could be easier to understand. As reference, I believe other solvers let you pass 2 arrays: a list of column indexes and a list of values for the variables.

**Reply by jajhall** [reply]
Yes, set them to any fractional value - it might be that an infeasible value is OK, actually. 

I just look for the values that are within the integrality tolerance of an integer value. I fix their bounds to that value, and (currently) solve the resulting MIP. In practice I should just run the MIP solver until it has found an integer solution. The only danger with this approach is that fixing the user's chosen values might make the problem infeasible, and that could take ages to identify. 

Interesting to learn about other solvers. That is a better way, even if it needs another `setSolution` method.

**Reply by guifcoelho** [reply]
[This is how Xpress does it](https://www.fico.com/fico-xpress-optimization/docs/latest/solver/optimizer/python/HTML/problem.addmipsol.html)

I recommend writing a new method for this. Maybe keep `setSolution` the way it is and add a not method called `setMipSolution` or `setHotStartSolution`. In my opinion, this topic is relevent enough to get its own interface method but that's your call.

Also, would you mind opening this discussion again so it could be easier to find?

**Reply by jajhall** [reply]
Thanks. I hadn't meant to close the discussion. I'll reopen it once I've walked the dog
